src/Algebra/Graph/Undirected.hs

```diff
@@ -0,0 +1,913 @@
+{-# INLINE star #-}
+
+-- | The /stars/ formed by overlaying a list of 'star's. An inverse of
+-- 'adjacencyList'.
+-- Complexity: /O(L)/ time, memory and size, where /L/ is the total size of the
+-- input.
+--
+-- @
+-- stars []                      == 'empty'
+-- stars [(x, [])]               == 'vertex' x
+-- stars [(x, [y])]              == 'edge' x y
+-- stars [(x, ys)]               == 'star' x ys
+-- stars                         == 'overlays' . 'map' ('uncurry' 'star')
+-- stars . 'adjacencyList'         == id
+-- 'overlay' (stars xs) (stars ys) == stars (xs ++ ys)
+-- @
+stars :: [(a, [a])] -> Graph a
+stars = coerce1 G.stars
+{-# INLINE stars #-}
+
+-- | The /tree graph/ constructed from a given 'Tree.Tree' data structure.
+-- Complexity: /O(T)/ time, memory and size, where /T/ is the size of the
+-- given tree (i.e. the number of vertices in the tree).
+--
+-- @
+-- tree (Node x [])                                         == 'vertex' x
+-- tree (Node x [Node y [Node z []]])                       == 'path' [x,y,z]
+-- tree (Node x [Node y [], Node z []])                     == 'star' x [y,z]
+-- tree (Node 1 [Node 2 [], Node 3 [Node 4 [], Node 5 []]]) == 'edges' [(1,2), (1,3), (3,4), (3,5)]
+-- @
+tree :: Tree.Tree a -> Graph a
+tree = coerce1 G.tree
+
+-- | The /forest graph/ constructed from a given 'Tree.Forest' data structure.
+-- Complexity: /O(F)/ time, memory and size, where /F/ is the size of the
+-- given forest (i.e. the number of vertices in the forest).
+--
+-- @
+-- forest []                                                  == 'empty'
+-- forest [x]                                                 == 'tree' x
+-- forest [Node 1 [Node 2 [], Node 3 []], Node 4 [Node 5 []]] == 'edges' [(1,2), (1,3), (4,5)]
+-- forest                                                     == 'overlays' . 'map' 'tree'
+-- @
+forest :: Tree.Forest a -> Graph a
+forest = coerce1 G.forest
+
+-- | Remove a vertex from a given graph.
+-- Complexity: /O(s)/ time, memory and size.
+--
+-- @
+-- removeVertex x ('vertex' x)       == 'empty'
+-- removeVertex 1 ('vertex' 2)       == 'vertex' 2
+-- removeVertex x ('edge' x x)       == 'empty'
+-- removeVertex 1 ('edge' 1 2)       == 'vertex' 2
+-- removeVertex x . removeVertex x == removeVertex x
+-- @
+removeVertex :: Eq a => a -> Graph a -> Graph a
+removeVertex = coerce4 G.removeVertex
+{-# SPECIALISE removeVertex :: Int -> Graph Int ->
+  Graph Int #-}
+
+-- | Remove an edge from a given graph.
+-- Complexity: /O(s)/ time, memory and size.
+--
+-- @
+-- removeEdge x y ('edge' x y)       == 'vertices' [x,y]
+-- removeEdge x y . removeEdge x y == removeEdge x y
+-- removeEdge x y                  == removeEdge y x
+-- removeEdge x y . 'removeVertex' x == 'removeVertex' x
+-- removeEdge 1 1 (1 * 1 * 2 * 2)  == 1 * 2 * 2
+-- removeEdge 1 2 (1 * 1 * 2 * 2)  == 1 * 1 + 2 * 2
+-- @
+removeEdge :: Eq a => a -> a -> Graph a -> Graph a
+removeEdge s t = coerce (G.removeEdge s t . G.removeEdge t s) 
```

Comment:
What does `coerce` do in this case? I think it should be removed.

Reply:
`coerce` makes the `G.removeEdge s t . G.removeEdge t s` which is of type `G.Graph a -> G.Graph a`, to be of type `Graph a -> Graph a`, right?

I can change the function and not use `coerce` with explicit pattern matching!

Reply:
@snowleopard forgot to mention you in the comment above :-)

Reply:
Thanks, I think I got it, let's keep it as is.

P.S.: No need to mention me, I get notifications either way :) It's just a very busy week.

Reply:
Thanks, and sorry I forgot it was a review! I didn't mean to be impatient! Take your time!